### ISSUE TITLE
Remove Conversion#register()

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-engine#1640]
Extracted from #1639.

What's the point of `Conversion#register()` if `for()` returns only `UpcastConversionsHelpers` or `DowncastConversionHelpers` (according to its docs)? In fact, I'd love if we could somehow remove `register()` because realistically speaking it can be used together with `new Conversion()` only (just like it currently is). You can't register your own conversion helpers. We could make it possible to pass all helpers directly to `new Conversion()`. It's not critical, but I found it a bit weird and confusing how it's documented right now.

**Comment by jodator** [comment]
So the `register()` method is used to define the "upcast", "downcast", "editingDowncast" and "dataDowncast" dispatchers groups.

The Conversion on its own requires that "upcast" and "downcast" groups are defined (used in two-way-converters like `editor.conversion.elmentToElement()`) so this makes `new Conversion()` a broken instance. To make that beautiful we _could_ introduce some builder:

```js
const builder=  new ConversionBuilder();
builder.addDispatcher( 'downcast', this.data.downcastDispatcher, 'dataDowncast' );
builder.addDispatcher( 'downcast', this.editing.downcastDispatcher, 'editingDowncast' );
builder.addDispatcher( 'upcast', this.data.upcastDispatcher );

// This could check for required 'upcast'/'downcast'
this.conversion = builder.build();
```
but this too Java-ish probably.

Note: in above methods you only define dispatcher once with optional alias (ie for `dataDowncast`).

The other thing is we could leave this as is and refactor logic a bit:
```
this.conversion = new Conversion();

this.conversion.addDispatcher( 'downcast', this.data.downcastDispatcher, 'dataDowncast' );
this.conversion.addDispatcher( 'downcast', this.editing.downcastDispatcher, 'editingDowncast' );
this.conversion.addDispatcher( 'upcast', this.data.upcastDispatcher );
```

or even more in line with the requirement for `upcast`/`downcast`:

```js
this.conversion.addDowncastDispatcher( this.data.downcastDispatcher, 'dataDowncast' );
this.conversion.addDowncastDispatcher( this.editing.downcastDispatcher, 'editingDowncast' );
this.conversion.addUpcastDispatcher( this.data.upcastDispatcher );
```

as we require `upcast`/`downcast` dispatchers to be available in `Conversion` anyway.

Another solution would be pass everyghint to the constructor:

```js
this.conversion = new Conversion( {
	upcast: this.data.upcastDispatcher,
	downcast: {
		dataDowncast: this.data.downcastDispatcher,
		editingDowncast: this.editing.dataDowncast
	}
} );
```
but is also ugly.

---

tl;dr - we cannot just remove `.register()` method. We must add some new API and I'd go with:

```js
this.conversion = new Conversion();

this.conversion.addDowncastDispatcher( this.data.downcastDispatcher, 'dataDowncast' );
this.conversion.addDowncastDispatcher( this.editing.downcastDispatcher, 'editingDowncast' );
this.conversion.addUpcastDispatcher( this.data.upcastDispatcher );
```

cc @pjasiun & @Reinmar, @scofalik (?)

**Comment by pjasiun** [comment]
Well... for me what we have now (`register` is fine, so its up to @Reinmar who was against it).

But some more ideas we had together with @jodator:

Simple names - values object:
```js
this.conversion = new Conversion( {
	'downcast': new DowncastHelpers( [ this.editing.downcastDispatcher, this.data.downcastDispatcher ] ),
	'editingDowncast': new DowncastHelpers( this.editing.downcastDispatcher ),
	'dataDowncast', new DowncastHelpers( this.data.downcastDispatcher ),
	'upcast': new UpcastHelpers( this.data.upcastDispatcher )
} );
```
2 parameters downcast helpers and then upcast helpers (`new DowncastHelpers` and ` new UpcastHelpers` can be omit):
```js
this.conversion = new Conversion( {
		'downcast': [ this.editing.downcastDispatcher, this.data.downcastDispatcher ],
		'editingDowncast': this.editing.downcastDispatcher,
		'dataDowncast', this.data.downcastDispatcher
	}, {
		'upcast': this.data.upcastDispatcher
	} );
```

`downcast` and `upcast` as default keys:
```js
this.conversion = new Conversion( {
		'editingDowncast': this.editing.downcastDispatcher,
		'dataDowncast', this.data.downcastDispatcher
	}, this.data.upcastDispatcher );
```

Downcast, upcast, then aliases:
```js
this.conversion = new Conversion( [ this.editing.downcastDispatcher, this.data.downcastDispatcher ], this.data.upcastDispatcher, {
	'editingDowncast': this.editing.downcastDispatcher,
	'dataDowncast', this.data.downcastDispatcher 
} );
```
`addAlias` as a separate methods:
```js
this.conversion = new Conversion( [ this.editing.downcastDispatcher, this.data.downcastDispatcher ], this.data.upcastDispatcher );
this.conversion.addAlias( 'editingDowncast', this.editing.downcastDispatcher );
this.conversion.addAlias( 'dataDowncast', this.data.downcastDispatcher  );
```

**Comment by pjasiun** [comment]
And I think that the last proposal (`addAlias` as a separate methods) is the best.
